middle_end/flambda2/terms/set_of_closures.ml

```diff
@@ -102,9 +102,13 @@ let [@ocamlformat "disable"] print ppf
       (Var_within_closure.Map.print Simple.print) closure_elements
 
 let free_names { function_decls; closure_elements } =
-  Name_occurrences.union_list
-    [ Function_declarations.free_names function_decls;
-      Simple.List.free_names (Var_within_closure.Map.data closure_elements) ]
+  Var_within_closure.Map.fold
+    (fun closure_var bound_to free_names ->
+      Name_occurrences.add_closure_var
+        (Name_occurrences.union (Simple.free_names bound_to) free_names)
+        closure_var Name_mode.normal)
+    closure_elements
+    (Function_declarations.free_names function_decls)
```

Comment:
You're changing the meaning of `Name_occurrences.closure_vars` by also registering those in sets of closures (instead of only projections).
I'm nor sure it's harmless: the only place during `Simplify` where I could see unused closure vars removed is in `Expr_builder.remove_unused_closure_vars`, and as far as I can tell this is only used on lifted sets of closures. So `To_cmm` will now keep all closure variables present in dynamically allocated sets of closures.

Reply:
You're right, I'll revert this code to its current state in `main`, so that the free_names of a set of closures do not include the closure_ids / var_within_closures that it binds.